File: csharp/1583.count-unhappy-friends.cs
/*
 * @lc app=leetcode id=1583 lang=csharp
 *
 * [1583] Count Unhappy Friends
 */

// @lc code=start
public class Solution {
    public int UnhappyFriends(int n, int[][] preferences, int[][] pairs) {
        int[] map = new int[n];
        int[,] pmap = new int[n, n];
        int ret = 0;
        foreach (var pair in pairs) {
            map[pair[0]] = pair[1];
            map[pair[1]] = pair[0];
        }
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n - 1; ++j) {
                pmap[i, preferences[i][j]] = j;
            }
        }
        for (int i = 0; i < n; ++i) {
            int p = map[i];
            for (int j = 0; j < n; ++j) {
                if (j == i || j == p) {
                    continue;
                }
                int jp = map[j];
                if (pmap[i, j] < pmap[i, p] && pmap[j, i] < pmap[j, jp]) {
                    ++ret;
                    break;
                }
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/1583.count-unhappy-friends.java
/*
 * @lc app=leetcode id=1583 lang=java
 *
 * [1583] Count Unhappy Friends
 */

// @lc code=start
class Solution {
    public int unhappyFriends(int n, int[][] preferences, int[][] pairs) {
        int ret = 0;
        int[] friends = new int[n];
        int[][] prefs = new int[n][n];
        for (int[] pair : pairs) {
            friends[pair[0]] = pair[1];
            friends[pair[1]] = pair[0];
        }
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n - 1; ++j) {
                prefs[i][preferences[i][j]] = j;
            }
        }
        for (int i = 0; i < n; ++i) {
            int f = friends[i];
            for (int j = 0; j < n; ++j) {
                if (i == j || f == j) {
                    continue;
                }
                if (prefs[i][j] < prefs[i][f] && prefs[j][i] < prefs[j][friends[j]]) {
                    ++ret;
                    break;
                }
            }
        }
        return ret;
    }
}
// @lc code=end

